### PR TITLE
Add unit tests for pkg/etcd

### PR DESCRIPTION
**Pull request by dereknola**
#### Proposed Changes ####
This is the first of hopefully many new unit tests. 
This change adds a new `unit_tests` package that contains a `runtime.go` file which provides tests with a default `config.ControlRuntime` object, filled out with the default certificates. It also allows unit tests to create folder structures under `/tmp/k3s` that duplicates the standard layout in `/var/lib/rancher/k3s`. 

It also adds a `etcd_test.go` file that sets several functions in the `pkg/etcd/etcd.go` file. 

#### Types of Changes ####
Unit Tests

#### Verification ####
Run `make test-unit` and see the new unit tests run.

#### Linked Issues ####
https://github.com/k3s-io/k3s/issues/3496

#### Further Comments ####
These tests were autogenerated using the `gotests` tool, with a modified template (which should be available soon<sup>TM</sup>). No external testing libraries are used. This PR is not just about the tests themselves, but the methodology and layout to guide future unit tests, so feel free to comment on that as well. 

I do not want the `unit_tests` package to become large. However, config.Control Config and config.Runtime are used throughout the codebase and will be needed for more tests. Thus having a central package made sense.

### CODE REVIEW

File: pkg/unit_tests/runtime.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
+package unit_tests
```
Comment (briandowns) [inline pkg/unit_tests/runtime.go:1]: Convention is to not have any form of special characters in package names. But, we have a utils package already. Might make sense to create a package in there called "test" and place this file in there. We can rename the functions and remove the "test" portion from the name sine the imported reference will provide that context.

File: pkg/etcd/etcd_test.go [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+	"testing"
+
+	"github.com/rancher/k3s/pkg/clientaccess"
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/k3s/pkg/unit_tests"
+	"github.com/robfig/cron/v3"
+	etcd "go.etcd.io/etcd/clientv3"
+)
+
+type Setup func() error
```
Comment (briandowns) [inline pkg/etcd/etcd_test.go:17]: Is this type used in another package? If not, this and the type definition below should be unexported.

File: pkg/etcd/etcd_test.go [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+		EtcdS3Endpoint:        "s3.amazonaws.com",
+		EtcdS3Region:          "us-east-1",
+		SANs:                  []string{"127.0.0.1"},
+	}
+	return &defaultTestConfig
+}
+
+func generateTestHandler() http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+	})
```
Comment (briandowns) [inline pkg/etcd/etcd_test.go:39]: Nit: move this up to be inline.

File: pkg/etcd/etcd_test.go [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+				return nil
+			},
+			wantErr: false,
+			want:    false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			e := NewETCD()
+			defer tt.teardown(tt.args.config)
```
Comment (briandowns) [inline pkg/etcd/etcd_test.go:101]: Awesome.

File: pkg/unit_tests/runtime.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	"net"
+	"os"
+	"path/filepath"
+
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/k3s/pkg/daemons/control/deps"
+	"github.com/rancher/k3s/pkg/version"
+)
+
+// All certs and other files are stored in /tmp/k3s/<RANDOM_STRING>/
```
Comment (briandowns) [inline pkg/unit_tests/runtime.go:13]: Can you update the doc comments to be in the expected format? https://golang.org/doc/effective_go#commentary

File: pkg/unit_tests/runtime.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	}
+	return nil
+}
+
+func CleanupTestDataDir(cnf *config.Control) {
+	os.RemoveAll(cnf.DataDir)
+}
+
+func GenerateTestRuntime(cnf *config.Control) error {
+	var err error
```
Comment (briandowns) [inline pkg/unit_tests/runtime.go:42]: This can be removed and an colon added on line 54.

File: pkg/etcd/etcd_test.go [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+	"github.com/rancher/k3s/pkg/unit_tests"
+	"github.com/robfig/cron/v3"
+	etcd "go.etcd.io/etcd/clientv3"
+)
+
+type Setup func() error
+type Teardown func() error
+
+func generateTestConfig() *config.Control {
+	var defaultTestConfig = config.Control{
```
Comment (briandowns) [inline pkg/etcd/etcd_test.go:21]:
```suggestion
	return &config.Control{
```

File: pkg/etcd/etcd_test.go (hunk 2 above)
Comment (briandowns) [reply pkg/etcd/etcd_test.go:17]: I wonder if it'd be possible to define these in the package I mention below and have the generator reference them from there.

File: pkg/etcd/etcd_test.go [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+			name: "Directory exists",
+			args: args{
+				ctx:    context.TODO(),
+				config: generateTestConfig(),
+			},
+			setup: func(cnf *config.Control) error {
+				if err := unit_tests.GenerateTestDataDir(cnf); err != nil {
+					return err
+				}
+				err := os.MkdirAll(walDir(cnf), 0700)
```
Comment (briandowns) [inline pkg/etcd/etcd_test.go:65]:
```suggestion
				return os.MkdirAll(walDir(cnf), 0700)
```

File: pkg/etcd/etcd_test.go [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+			},
+			setup: func(cnf *config.Control) error {
+				if err := unit_tests.GenerateTestRuntime(cnf); err != nil {
+					return err
+				}
+				if err := os.MkdirAll(etcdDBDir(cnf), 0700); err != nil {
+					return err
+				}
+				tombstoneFile := filepath.Join(etcdDBDir(cnf), "tombstone")
+				if _, err := os.Create(tombstoneFile); err != nil {
```
Comment (briandowns) [inline pkg/etcd/etcd_test.go:163]:
```suggestion
				return os.Create(tombstoneFile); err != nil {
```

File: pkg/etcd/etcd_test.go [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+	}{
+		{
+			name: "Call Register with standard config",
+			args: args{
+				ctx:     context.TODO(),
+				config:  generateTestConfig(),
+				handler: generateTestHandler(),
+			},
+			setup: func(cnf *config.Control) error {
+				err := unit_tests.GenerateTestRuntime(cnf)
```
Comment (briandowns) [inline pkg/etcd/etcd_test.go:140]:
```suggestion
				return unit_tests.GenerateTestRuntime(cnf)
```

File: pkg/etcd/etcd_test.go [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+				config:  generateTestConfig(),
+				address: "192.168.1.123", // Local IP address
+			},
+			args: args{
+				ctx:              context.TODO(),
+				clientAccessInfo: nil,
+			},
+			setup: func(cnf *config.Control) error {
+				err := unit_tests.GenerateTestRuntime(cnf)
+				cnf.EtcdDisableSnapshots = true
```
Comment (briandowns) [inline pkg/etcd/etcd_test.go:228]: I think this should go on the line above and `return unit_tests.GenerateTestRuntime(cnf)` since I'm not sure you want to continue if there's an error generating the runtime.

File: pkg/etcd/etcd_test.go [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+				config:  generateTestConfig(),
+				address: "192.168.1.123", // Local IP address
+				cron:    cron.New(),
+			},
+			args: args{
+				ctx:              context.TODO(),
+				clientAccessInfo: nil,
+			},
+			setup: func(cnf *config.Control) error {
+				err := unit_tests.GenerateTestRuntime(cnf)
```
Comment (briandowns) [inline pkg/etcd/etcd_test.go:248]:
```suggestion
				return unit_tests.GenerateTestRuntime(cnf)
```

File: pkg/unit_tests/runtime.go [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/k3s/pkg/daemons/control/deps"
+	"github.com/rancher/k3s/pkg/version"
+)
+
+// All certs and other files are stored in /tmp/k3s/<RANDOM_STRING>/
+// for testing
+func GenerateTestDataDir(cnf *config.Control) error {
+	var err error
```
Comment (briandowns) [inline pkg/unit_tests/runtime.go:16]: This can be removed and an colon added on line 17.

File: pkg/unit_tests/runtime.go (hunk 1 above)
Comment (dereknola) [reply pkg/unit_tests/runtime.go]: I have renamed the package to "tests"

File: pkg/etcd/etcd_test.go (hunk 2 above)
Comment (dereknola) [reply pkg/etcd/etcd_test.go]: These are actually not needed and are detrimental to the test infra, because each test cannot redefine their setup/teardown function arguments by using these type defs. They have been removed.

File: pkg/etcd/etcd_test.go (hunk 3 above)
Comment (dereknola) [reply pkg/etcd/etcd_test.go]: Moved

File: pkg/unit_tests/runtime.go (hunk 5 above)
Comment (dereknola) [reply pkg/unit_tests/runtime.go]: I have updated all function comments to conform to the expected format

File: pkg/etcd/etcd_test.go (hunk 7 above)
Comment (dereknola) [reply pkg/etcd/etcd_test.go]: Commited

File: pkg/etcd/etcd_test.go (hunk 9 above)
Comment (dereknola) [reply pkg/etcd/etcd_test.go:155]: In this and other areas with the same format I have changed to the suggested "return FUNC" format

Comment (briandowns) [reply pkg/etcd/etcd_test.go:155]: I made this comment in error. Disregard.

File: pkg/unit_tests/runtime.go (hunk 13 above)
Comment (dereknola) [reply pkg/unit_tests/runtime.go]: Removed any  "var err error" and just changed the first time to "err :="

File: pkg/etcd/etcd_test.go (hunk 12 above)
Comment (dereknola) [reply pkg/etcd/etcd_test.go]: Fixed

File: pkg/etcd/etcd_test.go (hunk 11 above)
Comment (dereknola) [reply pkg/etcd/etcd_test.go:219]: Fixed, point is valid

File: pkg/etcd/etcd_test.go (hunk 8 above)
Comment (dereknola) [reply pkg/etcd/etcd_test.go]: Fixed

File: pkg/unit_tests/runtime.go (hunk 6 above)
Comment (dereknola) [reply pkg/unit_tests/runtime.go]: Fixed

File: pkg/unit_tests/runtime.go (hunk 1 above)
Comment (dereknola) [reply pkg/unit_tests/runtime.go]: Additionally, I have moved the tests package under "pkg/util"

File: pkg/etcd/etcd_test.go [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+		EtcdSnapshotName:      "etcd-snapshot",
+		EtcdSnapshotCron:      "0 */12 * * *",
+		EtcdSnapshotRetention: 5,
+		EtcdS3Endpoint:        "s3.amazonaws.com",
+		EtcdS3Region:          "us-east-1",
+		SANs:                  []string{"127.0.0.1"},
+	}
+}
+
+func generateTestHandler() http.Handler {
```
Comment (briandowns) [inline pkg/etcd/etcd_test.go:33]: Would this also be a good candidate to be put in the test package for use elsewhere?

File: pkg/util/tests/runtime.go [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/k3s/pkg/daemons/control/deps"
+	"github.com/rancher/k3s/pkg/version"
+)
+
+// GenerateTestDataDir creates a temporary directory at "/tmp/k3s/<RANDOM_STRING>/".
+// The latest directory created with this function is soft linked to "/tmp/k3s/latest/".
+// This allows tests to replicate the "/var/lib/rancher/k3s" directory structure
+func GenerateTestDataDir(cnf *config.Control) error {
```
Comment (briandowns) [inline pkg/util/tests/runtime.go:16]: Can you remove the "Test" portion from these funtion names? It's redudant as the package ref when used will provide context. ie `test.GenerateDataDir`.

File: pkg/util/tests/runtime.go [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+	"path/filepath"
+
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/k3s/pkg/daemons/control/deps"
+	"github.com/rancher/k3s/pkg/version"
+)
+
+// GenerateTestDataDir creates a temporary directory at "/tmp/k3s/<RANDOM_STRING>/".
+// The latest directory created with this function is soft linked to "/tmp/k3s/latest/".
+// This allows tests to replicate the "/var/lib/rancher/k3s" directory structure
```
Comment (briandowns) [inline pkg/util/tests/runtime.go:15]:
```suggestion
// This allows tests to replicate the "/var/lib/rancher/k3s" directory structure.
```

File: pkg/util/tests/runtime.go [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+// CleanupTestDataDir removes the associated "/tmp/k3s/<RANDOM_STRING>"
+// directory.
+func CleanupTestDataDir(cnf *config.Control) {
+	os.RemoveAll(cnf.DataDir)
+}
+
+// GenerateTestRuntime creates a temporary data dir and configures
+// config.ControlRuntime with all the appropriate certificate keys.
+func GenerateTestRuntime(cnf *config.Control) error {
+
```
Comment (briandowns) [inline pkg/util/tests/runtime.go:46]: Nit: kill this new line.

File: pkg/util/tests/runtime.go [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+	runtime := &config.ControlRuntime{}
+	if err := GenerateTestDataDir(cnf); err != nil {
+		return err
+	}
+
+	os.MkdirAll(filepath.Join(cnf.DataDir, "tls"), 0700)
+	os.MkdirAll(filepath.Join(cnf.DataDir, "cred"), 0700)
+
+	runtime.ClientCA = filepath.Join(cnf.DataDir, "tls", "client-ca.crt")
+	runtime.ClientCAKey = filepath.Join(cnf.DataDir, "tls", "client-ca.key")
```
Comment (brandond) [inline pkg/util/tests/runtime.go:65]: This all looks like it was copied from https://github.com/k3s-io/k3s/blob/f5fbb9a9a84dd758cfa2212167229e657e18c0c5/pkg/daemons/control/server.go#L234 - should we perhaps just move that into a library function?

File: pkg/etcd/etcd_test.go [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+				cnf.EtcdDisableSnapshots = true
+				return tests.GenerateTestRuntime(cnf)
+			},
+			teardown: func(cnf *config.Control) error {
+				tests.CleanupTestDataDir(cnf)
+				return nil
+			},
+		},
+		{
+			name: "Start etcd without clientAccesInfo on",
```
Comment (brandond) [inline pkg/etcd/etcd_test.go:228]: nit
```suggestion
			name: "Start etcd without clientAccessInfo on",
```

File: pkg/util/tests/runtime.go (hunk 18 above)
Comment (dereknola) [reply pkg/util/tests/runtime.go:65]: There are several things we do not want to happen in the testing infra, such as boostraping and starting the cluster. However, I will move all the `filepath.join` lines into a seperate function that can be used by both server.go and runtime.go.

File: pkg/etcd/etcd_test.go (hunk 14 above)
Comment (dereknola) [reply pkg/etcd/etcd_test.go:33]: I did not feel like there was enough other functions in the pkg that required a handler as an argument. Thus I did not move it into runtime.go to ensure that the tests pkg was a slim as possible.

Comment (briandowns) [reply pkg/etcd/etcd_test.go:33]: We can leave here for now until we see how if it'd be useful elsewhere and in the RKE2 codebase.

File: pkg/util/tests/runtime.go (hunk 18 above)
Comment (brandond) [reply pkg/util/tests/runtime.go:65]: 👍🏻  on that. I think the mkdir and path building stuff is worth sharing; certainly don't need to actually boostrap the cluster.

File: pkg/etcd/etcd_test.go (hunk 19 above)
Comment (dereknola) [reply pkg/etcd/etcd_test.go]: Fixed

File: pkg/util/tests/runtime.go (hunk 17 above)
Comment (dereknola) [reply pkg/util/tests/runtime.go]: Killed

File: pkg/util/tests/runtime.go (hunk 16 above)
Comment (dereknola) [reply pkg/util/tests/runtime.go]: Fixed

File: pkg/util/tests/runtime.go (hunk 15 above)
Comment (dereknola) [reply pkg/util/tests/runtime.go]: Removed

File: pkg/daemons/control/deps/deps.go [hunk 20]
```diff
@@ -90,6 +90,67 @@ func KubeConfig(dest, url, caCert, clientCert, clientKey string) error {
 	return kubeconfigTemplate.Execute(output, &data)
 }
 
+// FillRuntimeCerts is responsible for filling out all the
+// .crt and .key filenames for a ControlRuntime
```
Comment (briandowns) [inline pkg/daemons/control/deps/deps.go:94]:
```suggestion
// .crt and .key filenames for a ControlRuntime.
```

Comment (dereknola) [reply pkg/daemons/control/deps/deps.go]: Fixed

**Review by briandowns — APPROVED** [review]
LGTM. This is an awesome PR.

**Review by brandond — APPROVED** [review]
